### PR TITLE
S2236 - DAO Drop-In functionality

### PR DESCRIPTION
**Pull request by SharonGoliath**

### CODE REVIEW

File: caom2utils/caom2utils/fits2caom2.py [hunk 1]
```diff
@@ -1037,7 +1041,203 @@ def get_configed_axes_count(self):
...
+            raise ValueError(
+                'Must set a value of {} for {}.'.format(to_enum_type.__name__,
+                                                        self.logging_name))
+        elif isinstance(value, to_enum_type):
+            return value
+        else:
+            return to_enum_type(value)
+
+
+class GenericParser(Parser):
```
Comment (andamian) [inline caom2utils/caom2utils/fits2caom2.py]: It seems to me that this class does the same thing as Parser. I would make a GenericParser the base non-abstract class that the FitsParser customizes. Postpone the creation of abstract Parser for the future only if needed by other types of parsers.

File: caom2utils/caom2utils/legacy.py [hunk 2]
```diff
@@ -78,7 +80,7 @@
 
 APP_NAME = 'fits2caom2'
 
-__all__ = ['main_app', 'update_blueprint']
+__all__ = ['main_app', 'update_blueprint', 'apply_java_config']
```
Comment (andamian) [inline caom2utils/caom2utils/legacy.py]: I'm not sure we need to expose any of this as they are not part of the library API.

File: caom2/caom2/diff.py [hunk 3]
```diff
@@ -202,33 +203,66 @@ def _get_sequence_differences(expected, actual, parent):
...
+        else:
+            match_found = False
+            tracking_report = None
+            for act_index, a in enumerate(actual):
+                temp_report = get_differences(e, a, label)
+                if temp_report is None:
+                    match_found = True
+                    actual_copy.remove(a)
+                    expected_copy.remove(e)
+                    break
```
[the remark this answers is not in the document]
Comment (SharonGoliath) [reply caom2/caom2/diff.py]: Ok.

File: caom2/caom2/diff.py [hunk 4]
```diff
@@ -202,33 +203,66 @@ def _get_sequence_differences(expected, actual, parent):
...
+        if isinstance(e, Chunk):
+            if len(expected) > 1:
+                report.append('Sequence:: more Chunks than expected {}'.format(
+                    len(expected)))
+            temp_report = get_differences(e, actual[0])
+            if temp_report is not None:
+                report.extend(temp_report)
+            actual_copy.pop()
+            expected_copy.pop()
+            break
```
[the remark this answers is not in the document]
Comment (SharonGoliath) [reply caom2/caom2/diff.py]: Ok.

File: caom2/caom2/diff.py [hunk 5]
```diff
@@ -202,33 +203,66 @@ def _get_sequence_differences(expected, actual, parent):
...
+                temp_report = get_differences(e, a, label)
+                if temp_report is None:
+                    match_found = True
+                    actual_copy.remove(a)
+                    expected_copy.remove(e)
+                    break
+                else:
+                    # pick the report with the least number of errors
+                    if tracking_report:
+                        if len(temp_report) < len(tracking_report):
```
[the remark this answers is not in the document]
Comment (SharonGoliath) [reply caom2/caom2/diff.py]: Ok.

File: caom2/caom2/tests/test_diffs.py [hunk 6]
```diff
@@ -108,3 +108,32 @@ def test_get_differences(self):
...
+        report = diff.get_differences(obs1, obs3, 'caom_test_instances')
+        assert len(report) == 1
+
+    def test_samples(self):
+        instances = caom_test_instances.Caom2TestInstances()
+        seq1 = instances.get_coord_axis1d()
+        seq2 = instances.get_coord_axis1d()
+
+        report = diff.get_differences(seq1, seq2, 'samples')
+        assert report is None
```
[the remark this answers is not in the document]
Comment (SharonGoliath) [reply caom2/caom2/tests/test_diffs.py]: Added modification test.

File: caom2/caom2/tests/test_diffs.py [hunk 7]
```diff
@@ -108,3 +108,32 @@ def test_get_differences(self):
...
+
+    def test_chunks(self):
+        instances = caom_test_instances.Caom2TestInstances()
+        seq1 = instances.get_chunks()
+        seq2 = instances.get_chunks()
+
+        report = diff.get_differences(seq1, seq2, 'chunks')
+        assert report is None
```
[the remark this answers is not in the document]
Comment (SharonGoliath) [reply caom2/caom2/tests/test_diffs.py]: Added modification test.

File: caom2utils/caom2utils/fits2caom2.py [hunk 8]
```diff
@@ -1037,7 +1041,203 @@ def get_configed_axes_count(self):
...
+
+    @abstractmethod
+    def augment_observation(self, observation, artifact_uri, product_id=None):
+        """
+        Augments a given observation with plane structure only.
+        :param observation: existing CAOM2 observation to be augmented.
+        :param artifact_uri: the key for finding the artifact to augment
+        :param product_id: the key for finding for the plane to augment
+        """
+        assert observation
```
[the remark this answers is not in the document]
Comment (SharonGoliath) [reply caom2utils/caom2utils/fits2caom2.py]: All asserts in legacy.py and fits2caom2.py have a message.

File: caom2utils/caom2utils/fits2caom2.py (hunk 1 above)
Comment (SharonGoliath) [reply caom2utils/caom2utils/fits2caom2.py]: Parser class removed, GenericParser a base non-abstract class.

File: caom2utils/caom2utils/legacy.py (hunk 2 above)
Comment (SharonGoliath) [reply caom2utils/caom2utils/legacy.py]: Removed.

**Pull request merged**